dali/python/backend_impl.cc

```diff
@@ -1248,34 +1248,46 @@ PYBIND11_MODULE(backend_impl, m) {
     .def("RunCPU", &Pipeline::RunCPU, py::call_guard<py::gil_scoped_release>())
     .def("RunGPU", &Pipeline::RunGPU)
     .def("Outputs",
-        [](Pipeline *p) {
+        [](py::object &p_o) {
+          Pipeline *p = py::cast<Pipeline *>(p_o);
           DeviceWorkspace ws;
           p->Outputs(&ws);
 
-          py::list list;
+          py::tuple outs(ws.NumOutput());
           for (int i = 0; i < ws.NumOutput(); ++i) {
             if (ws.OutputIsType<CPUBackend>(i)) {
-              list.append(&ws.Output<CPUBackend>(i));
+              outs[i] = &ws.Output<CPUBackend>(i);
             } else {
-              list.append(&ws.Output<GPUBackend>(i));
+              outs[i] = &ws.Output<GPUBackend>(i);
             }
+            // it is not possible to obtain weakref  to the tuple and bound it to the pipeline
+            // object to keep the pipeline object alive as long as a returned tuple is valid,
+            // so instead keep pipeline alive as long as any element of the returned
+            // tuple is alive
+            py::detail::keep_alive_impl(outs[i], p_o);
           }
-          return py::cast<py::tuple>(list);
+          return outs;
         }, py::return_value_policy::take_ownership)
     .def("ShareOutputs",
-        [](Pipeline *p) {
+        [](py::object &p_o) {
+          Pipeline *p = py::cast<Pipeline *>(p_o);
           DeviceWorkspace ws;
           p->ShareOutputs(&ws);
 
-          py::list list;
+          py::tuple outs(ws.NumOutput());
           for (int i = 0; i < ws.NumOutput(); ++i) {
             if (ws.OutputIsType<CPUBackend>(i)) {
-              list.append(&ws.Output<CPUBackend>(i));
+              outs[i] = &ws.Output<CPUBackend>(i);
             } else {
-              list.append(&ws.Output<GPUBackend>(i));
+              outs[i] = &ws.Output<GPUBackend>(i);
             }
+            // it is not possible to obtain weakref  to the tuple and bound it to the pipeline
```

Comment:
```suggestion
            // it is not possible to obtain weakref to the tuple and bind it to the pipeline
```
Same above.

Btw, why did you change it to tuple?

Reply:
We have returned it as tuple anyway `return  py::cast<py::tuple>(list);`.

Reply:
Done